### PR TITLE
Contributing guidelines and pre-commit hooks

### PR DESCRIPTION
**Pull request by haipinglu**
### Description
Major update to [contribution guidelines](https://github.com/pykale/pykale/blob/hplu/contributing-guidelines/CONTRIBUTING.md) and pre-commit hooks ready to be enforced.

- **Contribution guidelines have been updated and will guide our future developments. Please review it by [reading through it completely](https://github.com/pykale/pykale/blob/hplu/contributing-guidelines/CONTRIBUTING.md)**. Your feedback is very important. Imagine you are new to PyKale and see whether by reading this doc you know what and how to do for various involvements (except release and management). This doc will be very important for us to standardize our practices and for new comers to join this effort. 
- ReadMe has been lightly updated on how to contribute.
- Pre-commit hooks have been tested locally and configurations have been updated. The only hook not yet active is [`mypy`](https://mypy.readthedocs.io/en/stable/), which needs deeper changes to code that the code authors should do (to discuss).
- Auto-fix some style issues using pre-commit hooks. You do not have to review these changes. They are minor (mostly whitespace/ending problems) and I have randomly checked to be safe. See the [commit description](https://github.com/pykale/pykale/commit/cc83c1d135bd85bdfa2ea1131e1a0edb06a5a0cf) if you want to learn more.
  - The changes to requirements.txt are not ideal but we will revisit them later when discussing dependancies.
- Further manual fixes for remaining flake8 errors, all in the imports of `main.py` of `Examples`. Now all `sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))` in `Examples` have been removed. You should install (the latest) PyKale to use `kale API` in examples.
- Simplified PR template.

**Summary**: Major changes to **CONTRIBUTING.md** that need all to review carefully. Minor changes to README.md and others that you may skip if you like.

We will discuss any unresolved issues in our PyKale meeting this Thursday. We should all use pre-commit hooks after resolving all questions. Of course, you are welcome to try out first (see contributing guidelines). I did and it is beautiful.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).

### CODE REVIEW

**Comment by pz-white** [comment]
The contributing guideline is cool! A supplement about PR merge, should we set at least one reviewer approve for merging operation. As I see although the PR requests reviewers now, but I can directly merge without any limitation. I think a more secure measure that the mandatory peer review should be set and referred in guideline.

**Review by pz-white — APPROVED** [review]

**Comment by pz-white** [comment]
I think we can set main reviewers for different pykale components e.g. image for Ravio & Shuo and GraphML for me. If one PR needs to be merged to specific component, the approve from main reviewer should be necessary and other reviewers' are optional.

**Comment by haipinglu** [comment]
> The contributing guideline is cool! A supplement about PR merge, should we set at least one reviewer approve for merging operation. As I see although the PR requests reviewers now, but I can directly merge without any limitation. I think a more secure measure that the mandatory peer review should be set and referred in guideline.

Good point. I will look into such settings. No wonder I cannot find the auto-merge feature (though enabled).

**Review by sz144 — APPROVED** [review]
Good for me. Shall we consider to have some figures in readme/index (e.g. a flowchart of contributing) for general developers in future?

**Comment by sz144** [comment]
> 
> 
> I think we can set main reviewers for different pykale components e.g. image for Ravio & Shuo and GraphML for me. If one PR needs to be merged to specific component, the approve from main reviewer should be necessary and other reviewers' are optional.

I suggest we can have a rank of reviewers for each division of Pykale, and reviewers with different rankings have different level of review permissions, e.g. I can only review/approve coding style changes in GraphML. Is it too complicated?

**Comment by pz-white** [comment]
> > The contributing guideline is cool! A supplement about PR merge, should we set at least one reviewer approve for merging operation. As I see although the PR requests reviewers now, but I can directly merge without any limitation. I think a more secure measure that the mandatory peer review should be set and referred in guideline.
> 
> Good point. I will look into such settings. No wonder I cannot find the auto-merge feature (though enabled).

FYI, https://docs.github.com/en/github/administering-a-repository/managing-a-branch-protection-rule, the official doc about how to set code owner review for protected branch, which is very clear.

**Comment by haipinglu** [comment]
> > > The contributing guideline is cool! A supplement about PR merge, should we set at least one reviewer approve for merging operation. As I see although the PR requests reviewers now, but I can directly merge without any limitation. I think a more secure measure that the mandatory peer review should be set and referred in guideline.
> > 
> > 
> > Good point. I will look into such settings. No wonder I cannot find the auto-merge feature (though enabled).
> 
> FYI, https://docs.github.com/en/github/administering-a-repository/managing-a-branch-protection-rule, the official doc about how to set code owner review for protected branch.

Yes, that's right. I just created a rule on master, which also enabled the auto-merge feature. Thanks.

**Comment by haipinglu** [comment]
> > I think we can set main reviewers for different pykale components e.g. image for Ravio & Shuo and GraphML for me. If one PR needs to be merged to specific component, the approve from main reviewer should be necessary and other reviewers' are optional.
> 
> I suggest we can have a rank of reviewers for each division of Pykale, and reviewers with different rankings have different level of review permissions, e.g. I can only review/approve coding style changes in GraphML. Is it too complicated?

It seems that this can be done by defining [CODEOWNERS](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners).

**Comment by haipinglu** [comment]
> I think we can set main reviewers for different pykale components e.g. image for Ravio & Shuo and GraphML for me. If one PR needs to be merged to specific component, the approve from main reviewer should be necessary and other reviewers' are optional.

Yes, it seems possible to automate this by defining [CODEOWNERS](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners).

**Comment by haipinglu** [comment]
> Good for me. Shall we consider to have some figures in readme/index (e.g. a flowchart of contributing) for general developers in future?

Any good examples (links) for what's in your mind? 

MONAI is the top-quality one and I make it even more compact. Text is simple and compact, I think (so we may remove the figure in ReadMe later.)

The hyperlinks at the top of CONTRIBUTING.md serve the purpose of an index, is it clear there?

**Comment by haipinglu** [comment]
@pz-white @sz144 The idea of automating reviewer assignment is well motivated but how to achieve it may be a (big) problem. I have studied the use of [CODEOWNERS](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners). I think it does not fit us because of our unified API. No single member owns a single directory in the core API. On the other hand, our [trello board](https://trello.com/b/X8VBNAvf/pykale-api-overview) may be helpful in clarifying ownership, but so far, only I have provided input so the sustainability of update is questionable. 

To my understanding, to use [CODEOWNERS](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners), we have to define the owners for **every file** in our case, which is tedious and not scalable. 

Actually the automatically suggested reviewers are generally good, at least to my experience (again RecSys!!!). I will enforce at least one reviewer before any merge to master (very good point by Peizhen).

If you have better solutions, I am happy to discuss. I've created a card in the project for possible discussion on Thursday.

@XianyuanLiu @RaivoKoot Welcome your two cents.

**Comment by sz144** [comment]
> 
> 
> > Good for me. Shall we consider to have some figures in readme/index (e.g. a flowchart of contributing) for general developers in future?
> 
> Any good examples (links) for what's in your mind?
> 
> MONAI is the top-quality one and I make it even more compact. Text is simple and compact, I think (so we may remove the figure in ReadMe later.)
> 
> The hyperlinks at the top of CONTRIBUTING.md serve the purpose of an index, is it clear there?

The hyperlinks are clear for me. I do not have any good example I have seen, but the fowchart in our ReadMe.md reminded me we can have something similar in CONTRIBUTING.md yesterday.

**Review by XianyuanLiu — APPROVED** [review]
The contributing file is very helpful and useful and makes me more clear about pre-commit. Many thanks!

**Comment by XianyuanLiu** [comment]
If automating reviewer assignment is not available yet, is there a tool to require PR must get one fixed reviewer approvement and can we fix one reviewer as the top reviewer? The process is like this.

1. Someone starts PR with assigning it to the top reviewer and recommended area reviewers. Top and recommended area reviewer list and the assignment rules can be added into the PR template for notification.
2. Area reviewers comment and then approve. Area reviewers reject the PR without assigning to the top reviewer.
3. Top reviewer finds area reviewers' approvement and then merge. If nothing special, top reviewer only needs to check whether others finish the reviews and notice. This aims to minimize the top reviewer's work.

File: docs/requirements.txt
Comment (haipinglu) [reply docs/requirements.txt:21]: No. We need to confirm all dependencies, a topic to discuss in the next PyKale meeting: https://github.com/pykale/pykale/projects/9#card-52913368

File: CONTRIBUTING.md
Comment (haipinglu) [reply CONTRIBUTING.md:50]: Thanks. Fixed. In future, for such problems, you could directly correct and push.

File: CONTRIBUTING.md
Comment (haipinglu) [reply CONTRIBUTING.md:124]: Thanks. I will fix but in future, you could directly correct it.

File: CONTRIBUTING.md
Comment (haipinglu) [reply CONTRIBUTING.md:129]: Yes. Actually, I will add a note to the section that Reviewers could fix simple problems such as typos. This will make things move more efficiently.

File: CONTRIBUTING.md
Comment (haipinglu) [reply CONTRIBUTING.md:169]: Done and I will push now. Thanks.

**Comment by haipinglu** [comment]
> If automating reviewer assignment is not available yet, is there a tool to require PR must get one fixed reviewer approvement and can we fix one reviewer as the top reviewer? The process is like this.
> 
> 1. Someone starts PR with assigning it to the top reviewer and recommended area reviewers. Top and recommended area reviewer list and the assignment rules can be added into the PR template for notification.
> 2. Area reviewers comment and then approve. Area reviewers reject the PR without assigning to the top reviewer.
> 3. Top reviewer finds area reviewers' approvement and then merge. If nothing special, top reviewer only needs to check whether others finish the reviews and notice. This aims to minimize the top reviewer's work.

Sounds like the process of paper review, not sure whether this will be agile enough for software development. The top reviewer, unless he/she is full-time on PyKale, will become a bottleneck sooner or later, slowing things down. My availability over the year could be varying (e.g. teaching, key deadlines, holiday). Teamwork is key. 

We can discuss this on Thursday with this card: https://github.com/pykale/pykale/projects/9#card-53513007

**Review by RaivoKoot — APPROVED** [review]
Approve. The contributing guidelines are very detailed and descriptive. They are very good.

**Comment by bobturneruk** [comment]
Couple of thoughts on this - I've not looked at it in very great detail...

- Seems like this PR covers some off-topic changes to examples? May be better to avoid this to make PRs more focussed and easier to review.
- The pre-commit hooks (which run locally) have been configured to also run as a GitHub action on push / PR to `master`. It's possible that someone will push non-compliant changes having not installed the pre-commit hooks or had a problem running them. I suggest that we find the setting in GitHub to make passing the checks a condition of merging with `master`. This would traditionally be under [webhooks](https://docs.github.com/en/developers/webhooks-and-events/about-webhooks) but it may be different for GitHub actions. These checks could be evolved to run on multiple OSs / configurations and include automated testing.
- I tried to run the pre-commit hooks and saw this:
```
(pykale) PS C:\Users\Bob Turner\Documents\pykale> pre-commit run --all-files
[ERROR] The hook `isort` specifies `language_version` but is using language `system` which does not install an environment.  Perhaps you meant to use a specific language?
```
It's probably me not reading the instructions properly. I'll have another look when I get chance.

Great to see a team pushing for such high standards!

**Comment by haipinglu** [comment]
> Couple of thoughts on this - I've not looked at it in very great detail...

**Comment by haipinglu** [comment]
> Couple of thoughts on this - I've not looked at it in very great detail...
> 
> * Seems like this PR covers some off-topic changes to examples? May be better to avoid this to make PRs more focussed and easier to review.

Your suggestion is great and I will add it to the guidelines, though the changes to examples are not off-topic. The changes in `Examples` were made to address flake8 error `E402 module level import not at top of file`.
> * The pre-commit hooks (which run locally) have been configured to also run as a GitHub action on push / PR to `master`. It's possible that someone will push non-compliant changes having not installed the pre-commit hooks or had a problem running them. I suggest that we find the setting in GitHub to make passing the checks a condition of merging with `master`. This would traditionally be under [webhooks](https://docs.github.com/en/developers/webhooks-and-events/about-webhooks) but it may be different for GitHub actions. These checks could be evolved to run on multiple OSs / configurations and include automated testing.

The linting workflow (action) will run pre-commit hooks and report error for any failures. I have enforced *Require status checks to pass before merging* (to master) so a failure will prevent merging. I will test whether it does block merging when checks have failed. I also plan to rename the workflow `linting` to `check` to be more precise and also make it align better with pre-commit. Multi-OS checks will be good but we are following the convention `ubuntu-latest` in PyTorch and other libraries in its ecosystem to be consistent. Tests are not ready but once they are, integration is easy.
> * I tried to run the pre-commit hooks and saw this:
> 
> ```
> (pykale) PS C:\Users\Bob Turner\Documents\pykale> pre-commit run --all-files
> [ERROR] The hook `isort` specifies `language_version` but is using language `system` which does not install an environment.  Perhaps you meant to use a specific language?
> ```
> 
> It's probably me not reading the instructions properly. I'll have another look when I get chance.

I guess you did it before (23:40) last night. I found the same error last night (see that [action report](https://github.com/pykale/pykale/runs/1780486632?check_suite_focus=true)) and fixed it with [commit 5943386](https://github.com/pykale/pykale/pull/45/commits/59433863e170ffbc536762f8629d4abefc798787) `clean isort hook` above. That problem should be gone by now.
> Great to see a team pushing for such high standards!

Great review helping us improve it further. Many thanks Bob! I will proceed to test and make changes as above.

**Comment by bobturneruk** [comment]
Ah yes @haipinglu I just needed to `pull` and the error went.

On the CI - because the checks can actually change the code, I wonder if it will be possible to get a `pass` status, then commit the uncorrected code to `master` in the merge? This looks like a great approach, but I've not seen it done quite like this before!

**Comment by bobturneruk** [comment]
Looks like your checks are running!

**Comment by haipinglu** [comment]
> Ah yes @haipinglu I just needed to `pull` and the error went.
> 
> On the CI - because the checks can actually change the code, I wonder if it will be possible to get a `pass` status, then commit the uncorrected code to `master` in the merge? This looks like a great approach, but I've not seen it done quite like this before!

Local checks (on your machine) do change the code but remote CI checks by the workflow/action does not change the code, see [this example](https://github.com/pykale/pykale/runs/1712458857?check_suite_focus=true), which has frustrated me a lot. It ran black and "simulated" fixing the found problem but none file has been changed actually.

Therefore, when CI checks found a fixable problem and said "files are changed", it means files are changed on their system not on our repository. In this case, the CI will fail and merge cannot proceed. The problem has to be fixed locally.

**Comment by haipinglu** [comment]
> Looks like your checks are running!

All raised concerns have been addressed with new commits. Many thanks!

**Review by bobturneruk — APPROVED** [review]
LGTM

**Pull request merged**
